components/checkbox/CheckboxGroup.razor.cs

```diff
@@ -58,32 +114,85 @@ protected override void OnInitialized()
             _selectedValues ??= Array.Empty<string>();
         }
 
+        protected override void OnAfterRender(bool firstRender)
+        {
+            if (firstRender)
+            {
+                if (ChildContent is not null && _checkboxItems.Count > 0)
+                {
+                    if (Options.IsT0)
+                        _constructedOptions = _checkboxItems
+                            .Where(c => !c.IsFromOptions)
+                            .Select(c => new CheckboxOption { Label = c.Label, Value = c.Label, Checked = c.Value })
+                            .ToArray();
+                    else
+                    {
+                        _constructedOptions = _checkboxItems
+                            .Where(c => !c.IsFromOptions)
+                            .Select(c => c.Label).ToArray();
+                    }
+                }
+
+                if (ChildContent is not null && _isOptionDefined)
+                {
+                    Options.Match(
+                        opt => _currentValue = () => opt.Where(x => x.Checked).Select(x => x.Value)
+                                                    .Union(_constructedOptions.AsT0.Where(x => x.Checked).Select(x => x.Value))
+                                                    .ToArray(),
+                        arr => _currentValue = () => _selectedValues);
+                }
+                else
+                {
+                    var workWith = (_isOptionDefined ? Options : _constructedOptions);
+                    workWith.Match(
+                        opt => _currentValue = () => opt.Where(x => x.Checked).Select(x => x.Value).ToArray(),
+                        arr => _currentValue = () => _selectedValues);
+                }
```

Comment:
Could you please put these logic code into a method? There are many if-else here.

Reply:
Sure. I was planning to do that but rushed to get this out. I would have probably fixed it during test creation.